applications/tari_merge_mining_proxy/src/proxy.rs

```diff
@@ -0,0 +1,518 @@
+                    name: "NewBlockTemplate",
+                    details: err,
+                })?;
+        helpers::add_coinbase(&self.consensus, &mut new_block_template)?;
+        debug!(target: LOG_TARGET, "Added coinbase to new block template",);
+        let block = grpc_client
+            .get_new_block(grpc::NewBlockTemplate::from(new_block_template))
+            .await
+            .map_err(|status| MmProxyError::GrpcRequestError {
+                status,
+                details: "failed to get new block".to_string(),
+            })?
+            .into_inner();
+        debug!(
+            target: LOG_TARGET,
+            "Received new block from Tari base node #{}",
+            block
+                .block
+                .as_ref()
+                .and_then(|b| b.header.as_ref())
+                .map(|h| h.height)
+                .unwrap_or_default()
+        );
+        let mining_data = block
+            .clone()
+            .mining_data
+            .ok_or_else(|| MmProxyError::GrpcResponseMissingField("mining_data"))?;
+
+        let mut transient = self.state.transient_data.write().await;
+        transient.tari_block = Some(block);
+
+        // Deserialize the block template blob
+        let block_template_blob = &monerod_resp["result"]["blocktemplate_blob"];
+        let mut block = helpers::deserialize_from_hex::<_, blockdata::Block>(block_template_blob.to_string())?;
+
+        let input_blob = monero_rx::create_input_blob(&block)?;
+        monerod_resp["result"]["blockhashing_blob"] = input_blob.into();
+
+        // Add the Tari merge mining tag to the retrieved block template
+        monero_rx::append_merge_mining_tag(&mut block, mining_data.mergemining_hash.as_slice())?;
+        let mm_tag = helpers::serialize_to_hex(&block)?;
+        monerod_resp["result"]["blocktemplate_blob"] = mm_tag.into();
+
+        transient.monero_seed = Some(&monerod_resp["result"]["seed_hash"])
+            .filter(|v| !v.is_null())
+            .map(|v| v.to_string());
+
+        Ok(into_body(parts, monerod_resp))
+    }
+
+    async fn connect_grpc_client(
+        &self,
+    ) -> Result<grpc::base_node_client::BaseNodeClient<tonic::transport::Channel>, MmProxyError> {
+        let client =
+            grpc::base_node_client::BaseNodeClient::connect(format!("http://{}", self.config.grpc_address)).await?;
+        Ok(client)
+    }
+
+    fn get_fully_qualified_monerod_url(&self, uri: &Uri) -> Result<Url, MmProxyError> {
+        let uri = format!("{}{}", self.config.monerod_url, uri.path()).parse::<Url>()?;
+        Ok(uri)
+    }
+
+    /// Proxy a request received by this server to Monerod
+    async fn proxy_request_to_monerod(
+        &self,
+        mut req: Request<Body>,
+    ) -> Result<(Request<Bytes>, Response<json::Value>), MmProxyError>
+    {
+        let monerod_uri = self.get_fully_qualified_monerod_url(req.uri())?;
+        debug!(target: LOG_TARGET, "Proxying request to {}", monerod_uri);
```

Comment:
If the request is a post request, we would want to log the method in the request as well, since they all go to the same place (/json_rpc).

Reply:
Good idea!